[Feat] 카카오 소셜 로그인 및 JWT 인증 시스템 구현

supsup-hae opened the pull request:
## 💡 작업 내용

- [x] 인증 시스템 기본 구조 구현
- [x] OAuth2 + JWT 통합 인증 체계 구축
- [x] 카카오 소셜 로그인 연동
- [x] 토큰 기반 인증 필터 구현
- [x] 사용자 정보 처리 로직 개선
- [x] 로그아웃 구현

## 💡 자세한 설명

**필수 사항**
- [로컬 환경 카카오 로그인 검증 및 JWT 발급 테스트](http://localhost:8080/oauth2/authorization/kakao) (프로젝트 실행 후 테스트)
- .env 파일로 키를 관리했습니다. Notion - 백엔드 페이지에 첨부했으니 `application.yml`과 같은 위치에 위치해주세요! 

**OAuth2 설정 구현**
- Spring Security와 OAuth2 Client 의존성 추가
- 카카오 OAuth 클라이언트 설정 (client-id, secret, scope 등)
- OAuth2 인증 후 JWT 토큰 발급 플로우 구현
- OAuth2 로그인 성공/실패 핸들러 구현

**인증 처리 로직 개선**
- SecurityConfig를 통한 보안 설정 구성
- JWT 기반의 인증 필터 구현으로 토큰 검증 

**사용자 정보 관리**
- User 엔티티 설계 및 구현
- 소셜 로그인 사용자 정보 매핑 처리
- 토큰 재검증 및 갱신 로직 추가

## 📢 리뷰 요구 사항
- OAuth + JWT + Security 구현은 처음이라 미숙한 부분이 많을 것 같습니다 확인 부탁 드려요!
- 현재 `api/auth/logout` 은 미구현 상태입니다. 기능의 필요성 검토하고 문서 좀 더 읽어봐야 할 것 같습니다.

## ✅ 셀프 체크리스트
- [x] PR 제목을 형식에 맞게 작성했나요?
- [x] 브랜치 전략에 맞는 브랜치에 PR을 올리고 있나요?
- [x] Assignees, Reviewers, Labels 를 등록했나요?
- [x] 작업 도중 문서 수정이 필요한 경우 잘 수정했나요?
- [ ] 테스트는 잘 통과했나요?
- [x] 불필요한 코드는 제거했나요?

supsup-hae committed:
:wrench: chore: OAuth2, OpenAI, JWT 환경설정 추가 (#11)
- 카카오 OAuth2 인증 설정 및 의존성 추가
- OpenAI API 키 및 GPT 모델 설정 추가
- JWT 키 설정 및 의존성 추가

supsup-hae committed:
:recycle: refactor: 에러 처리 구조 개선 (#11)
- 에러 코드 패키지 구조 개선 (GlobalErrorCode, ErrorCode)
- 서비스 레이어 호환성 수정
- 관련 서비스 코드 리팩토링

supsup-hae committed:
sparkles: feat(entity): 사용자 및 토큰 도메인 구현 (#11)
- User 엔티티 클래스 수정
- Token 엔티티 클래스 구현

supsup-hae committed:
:sparkles: feat: 사용자 및 토큰 도메인 구현 (#11)
- User 엔티티 클래스 수정
- Token 엔티티 클래스 구현
Please enter the commit message for your changes. Lines starting

supsup-hae committed:
:sparkles: feat: 토큰 및 사용자 레포지토리 구현 (#11)
- TokenRepository 인터페이스 구현
- UserRepository 인터페이스 구현

supsup-hae committed:
:sparkles: feat: JWT 인증 필터 및 보안 설정 구현 (#11)
- SecurityConfig 보안 설정 구현
  - Spring Security 기본 설정
  - CORS, CSRF 설정
  - 인증/인가 예외 URL 설정
- JWT 인증 필터 구현
  - 토큰 검증 및 인증, 관련 예외 처리
- JWT 토큰 관리 클래스 구현
  - 토큰 생성/검증 로직 및 사용자 정보 추출

supsup-hae committed:
:sparkles: feat: OAuth2 인증 시스템 구현 (#11)
• 카카오 OAuth2 로그인 성공/실패 핸들러 구현
• OAuth2 사용자 정보 처리 및 변환 서비스 구현
• 카카오 프로필 정보 매핑 및 사용자 등록 처리

supsup-hae committed:
:sparkles: feat: 인증 컨트롤러 및 DTO 구현 (#11)
• 로그인/토큰 갱신 API 엔드포인트 구현
• 로그인 요청/응답 DTO 구현
• 토큰 응답 DTO 구현

supsup-hae committed:
:wrench: chore: 토큰 예외 처리 구현 (#11)
- 토큰 관련 예외 클래스 구현
- 토큰 예외 처리 로직 추가

supsup-hae committed:
:sparkles: feat: 사용자 UUID 생성기 구현 (#11)

supsup-hae committed:
:sparkles: feat: 토큰 서비스 구현 (#11)
- 토근 생성 및 삭제

supsup-hae committed:
Merge branch 'feature/11' of https://github.com/IT-Cotato/10th-ADIT-BE into feature/11

supsup-hae committed:
:sparkles: feat: 인증 오류 처리를 위한 커스텀 예외 추가 (#11)
- 인증 관련 예외를 체계적으로 관리하기 위한 AuthException 구현

supsup-hae committed:
:recycle: refactor: 토큰 관리 및 인증 로직 전면 개선 (#11)

- JWT 관련 토큰 분리 처리 구현
- 로그아웃 시 토큰 삭제 및 API 연동 개선
- JwtTokenProvider 에러 처리 강화

supsup-hae committed:
:fire: refactor: 사용하지 않는 로그인 관련 클래스 제거 (#11)
- 미사용 클래스 제거

supsup-hae committed:
:sparkles: feat: 토큰 관련 에러 코드 추가 및 접근 제어 개선 (#11)

supsup-hae committed:
:sparkles: feat: JWT 인증 및 Swagger 설정 확장 (#11)
- Swagger UI에 JWT 인증 기능 통합 및 보안 설정 개선

YoungjaeRo commented on src/main/java/com/adit/backend/global/error/GlobalErrorCode.java:
```diff
@@ -85,8 +86,13 @@ public enum GlobalErrorCode implements ErrorCode {
 
 	TOKEN_NOT_FOUND(UNAUTHORIZED, "1004", "토큰을 찾지 못했습니다."),
 
+	TOKEN_ALREADY_EXIST(UNAUTHORIZED, "1005", "토큰이 이미 생성되었습니다!"),
+
+	REFRESH_TOKEN_EXPIRED(UNAUTHORIZED,"1006" , "리프레쉬 토큰이 만료되었습니다."),
+
+	TOKEN_UNSURPPORTED(UNAUTHORIZED,"1007" ,"지원되지 않는 토큰입니다." ),
```
LGTM :)

YoungjaeRo commented on src/main/java/com/adit/backend/domain/auth/dto/OAuth2UserInfo.java:
```diff
@@ -0,0 +1,49 @@
+	String email,
+	String profile
+) {
+
+	public static OAuth2UserInfo of(String registrationId, Map<String, Object> attributes) throws AuthException {
+		return switch (registrationId) {
+			case "kakao" -> ofKakao(attributes);
+			default -> throw new AuthException(ILLEGAL_REGISTRATION_ID);
+		};
+	}
+
+	private static OAuth2UserInfo ofKakao(Map<String, Object> attributes) {
+		Map<String, Object> account = (Map<String, Object>)attributes.get("kakao_account");
+		Map<String, Object> profile = (Map<String, Object>)account.get("profile");
+
+		return OAuth2UserInfo.builder()
+			.name(String.valueOf(profile.get("nickname")))
+			.email(String.valueOf(account.get("email")))
+			.profile((String.valueOf(profile.get("profile_image_url"))))
+			.build();
+	}
+
+	public User toEntity() {
+		return User.builder()
+			.name(name)
+			.email(email)
+			.profile(profile)
+			.socialId(KeyGenerator.generateKey())
+			.nickname("test")
+			.role(Role.USER)
+			.build();
+	}
+}
```
고생하셨습니다! 간결하고 깔끔하게 구현되어, 읽기 편했습니다!

YoungjaeRo commented on src/main/java/com/adit/backend/domain/auth/dto/model/PrincipalDetails.java:
```diff
@@ -0,0 +1,64 @@
+
+import com.adit.backend.domain.user.entity.User;
+
+public record PrincipalDetails(
+	User user,
+	Map<String, Object> attributes,
+	String attributeKey) implements OAuth2User, UserDetails {
```
record 클래스는 이번에 처음 사용해보는거 같은데, 빨리 익숙해지겠습니다!

YoungjaeRo commented on src/main/java/com/adit/backend/global/security/oauth/CustomOAuth2UserService.java:
```diff
@@ -0,0 +1,55 @@
+	@Override
+	@Transactional
+	public OAuth2User loadUser(OAuth2UserRequest userRequest) throws OAuth2AuthenticationException {
+		// 유저 정보 가져오기
+		Map<String, Object> oAuth2UserAttributes = super.loadUser(userRequest).getAttributes();
+
+		// registrationId로 어떤 OAuth2 로그인인지 구분
+		String registrationId = userRequest.getClientRegistration().getRegistrationId();
+
+		String userNameAttributeName = userRequest.getClientRegistration().getProviderDetails()
+			.getUserInfoEndpoint().getUserNameAttributeName();
+
+		// OAuth2UserInfo 생성
+		OAuth2UserInfo oAuth2UserInfo = OAuth2UserInfo.of(registrationId, oAuth2UserAttributes);
+
+		// 유저 생성 또는 업데이트
+		User user = getOrSaveUser(oAuth2UserInfo);
+
+		return new PrincipalDetails(user, oAuth2UserAttributes, userNameAttributeName);
+	}
+
```
순차적으로 논리적인 로직으로 구현되어 리뷰하는 입장에서 많이 배웠습니다.

YoungjaeRo commented on src/main/java/com/adit/backend/global/security/oauth/handler/OAuth2SuccessHandler.java:
```diff
@@ -0,0 +1,39 @@
+			.build().toUriString();
+
+		response.sendRedirect(redirectUrl);
+	}
```
고생하셨습니다. LGTM

YoungjaeRo commented on src/main/java/com/adit/backend/domain/auth/entity/Token.java:
```diff
@@ -0,0 +1,52 @@
+	private String accessToken;
+
+	@Builder
+	public Token(User user, String refreshToken, String accessToken) {
+		this.user = user;
+		this.refreshToken = refreshToken;
+		this.accessToken = accessToken;
+	}
```
@AllArgsConstructor로 쉽게 구현해도 좋을거 같습니다!

supsup-hae replied:
그렇네요! 반영하겠습니다 :D

supsup-hae replied:
`@AllArgsConstructor`를 사용해 객체 생성시 ID 값도 포함해서 작성해야해서 불필요한 코드를 포함하고, 보안상 문제가 우려되었습니다! 그래서 `@RequiredArgsConstructor`을 명시한 점을 이용하여 엔티티에 직접 `@Builder` 를 사용해서 객체 생성을 하도록 변경했습니다.

supsup-hae committed:
:recycle: refactor: 토큰 생성에 빌더 패턴 적용 (#11)
- Token 엔티티에 빌더 패턴 적용
- TokenService 토큰 생성 로직 수정

supsup-hae committed:
:recycle: refactor: JWT 만료 시간 설정 추가 (#11)

supsup-hae committed:
:recycle: refactor: socialId 컬럼명 변경 및 Token 엔티티 개선 (#11)

supsup-hae committed:
:recycle: refactor: 토큰 처리 로직 개선 및 단순화 (#11)
- 액세스/리프레시 토큰 추출 프로세스 개선
- 토큰 설정 개선

supsup-hae committed:
:recycle: refactor: OAuth2 인증 로직 개선 및 로깅 강화 (#11)

- SecurityConfig 리팩터링
- OAuth2 인증 프로세스 개선

supsup-hae committed:
:recycle: refactor(auth): 토큰 관리 로직 개선 및 인증 흐름 정리 (#11)

- 토큰 관련 로직 재구성
  - DB 쿼리 최적화 및 캐싱 적용
  - 토큰 갱신 메커니즘 개선
- 인증 흐름 개선
  - OAuth 핸들러 로직 정리
  - 엔드포인트 구조 최적화
  - DTO 계층 구조 개선

supsup-hae committed:
:recycle: refactor: 프로젝트 구조 개선 (#11)

supsup-hae committed:
:sparkles: feat: 사용자 닉네임 정의 API 및 사용자 예외 처리 구현 (#11)

supsup-hae commented:
#### 12/17 회의 변경사항
- 기존의 시스템은 자체적으로 JWT 토큰을 발급하여 관리하는 목적이지만 프론트와 협업을 위해 통일된 응답을 보내는 것이 좋다.
- 따라서 카카오에서 발급하는 JWT 토큰을 이용 및 관리하는 것에 초점을 맞추어 개발을 진행하기로 했다.

supsup-hae committed:
:sparkles: feat: Kakao REST API를 이용한 토큰 발급 구현 완료 (#11)

supsup-hae committed:
:wrench: feat: JWT 필터 비활성화 및 Security 설정 변경 (#11)
- JWT 인증 필터 관련 컴포넌트 제거
- SecurityConfig 설정 재구성
- 기존 JWT 의존성 제거

supsup-hae committed:
:sparkles: feat: 카카오 로그인 및 토큰 관리 기능 구현 (#11)
- 카카오 REST API 연동하여 토큰 발급 로직 구현
- 토큰 기반 회원가입 및 로그인 프로세스 추가
- Token 엔티티 구조 개선 및 관련 DTO 생성
- 토큰 재발급 기능 구현

supsup-hae committed:
:sparkles: feat: 액세스 토큰 기반 닉네임 변경 로직 구현 중 (#11)
- 토큰 기반 회원 정보 수정 서비스 구현

supsup-hae committed:
:recycle: refactor: CQRS 패턴 적용을 위한 서비스 구조 개선 (#11)
- Command와 Query 서비스 분리 구현
- 인증 도메인 내 요청/응답 DTO 구조 개선
- 검증을 위한 예외처리 추가

supsup-hae committed:
:sparkles: feat: 토큰 Command 서비스 구현 및 적용 (#11)

supsup-hae committed:
:recycle: refactor: KakaoResponse Dto 구조 개선 (#11)

supsup-hae committed:
:sparkles: feat: 닉네임 변경 기능 구현 및 CQRS 패턴 적용 (#11)
- 닉네임 변경을 위한 Command/Query 서비스 분리
- Request/Response DTO 구조 개선
- DTO-Entity 변환을 위한 Converter 구현

supsup-hae committed:
:sparkles: feat(config): Swagger Authorization 헤더 설정 추가 (#11)
- Swagger UI에서 토큰 기반 API 테스트 가능

supsup-hae commented:
## 기능 구현 완료

![image](https://github.com/user-attachments/assets/fc297564-a970-4311-97d9-8468822aa529)
- 카카오 소셜 로그인을 이용한 Access, Refresh Token 발급 및 저장
- Token을 이용한 회원 기능 구현 (로그인, 로그아웃, 토큰 재발급, 회원 정보 조회, 닉네임 변경)
- [Swagger UI](http://localhost:8080/swagger-ui/index.html#/) 에서 토큰 기반 API 테스트 구현 - 로컬 환경에서 테스트

supsup-hae committed:
:recycle: refactor: 신규|기존 사용자 구분을 위한 응답형식 변경 (#11)
- 닉네임 변경시 Role이 GUEST -> USER로 변경
- 신규 사용자(Role.GUEST)
- 기존 사용자(Role.USER)

supsup-hae committed:
:sparkles: feat: 토큰 관리 방식 개선 및 보안 강화 (#11)
- 액세스 토큰: Authorization 헤더 | 리프레시 토큰 httpOnly 쿠키
- 토큰 관련 로깅 시스템 구현
- CORS 설정 추가 및 보안 강화
- 토큰 관리 코드 구조 개선 (로그인, 토큰 재발급)

supsup-hae committed:
:sparkles: feat: 회원 로그아웃 기능 구현 및 쿠키 제거 (#11)
- 로그아웃시 DB에서 Token 제거 및, Cookie 제거

supsup-hae commented:
### 프론트 요청 반영
![image](https://github.com/user-attachments/assets/cd50f135-154e-4ce4-9c16-16809c01a3d1)
![image](https://github.com/user-attachments/assets/517cc968-8cd0-4582-832f-06db202816ee)
- 카카오에서 발급한 `AccessToken -> Header` | `RefreshToken -> Cookie` 로 전달
- 신규 유저와, 기존 유저를 구분하기 위해 응답형식에 Role 추가 (`GUEST`: 신규, `USER`: 기존)

YoungjaeRo commented on build.gradle:
```diff
@@ -48,6 +50,10 @@ dependencies {
     // .env 파일 관리
     implementation 'me.paulschwarz:spring-dotenv:3.0.0'
 
+    // JWT
+    implementation 'io.jsonwebtoken:jjwt-api:0.12.3'
+    runtimeOnly 'io.jsonwebtoken:jjwt-impl:0.12.3'
+    runtimeOnly 'io.jsonwebtoken:jjwt-jackson:0.12.3'
```
Good :)

YoungjaeRo commented on src/main/java/com/adit/backend/domain/auth/controller/AuthController.java:
```diff
@@ -0,0 +1,70 @@
+public class AuthController {
+	public static final String ACCESS_TOKEN_HEADER = "Authorization";
+	private final AuthCommandService authCommandService;
+
+	@Operation(summary = "카카오 회원가입/로그인 요청")
+	@ApiResponses({
+		@io.swagger.v3.oas.annotations.responses.ApiResponse(responseCode = "302", description = "카카오 로그인 페이지로 Redirect"),
+	})
+	@GetMapping("/kakao")
+	public void kakaoLogin(HttpServletResponse response) throws IOException {
+		String kakaoAuthUrl = authCommandService.createKakaoAuthorizationUrl();
+		response.sendRedirect(kakaoAuthUrl);
+	}
+
+	@Operation(summary = "카카오 회원가입/로그인 결과", description = "request는 https로 요청되기 때문에 때문에 로컬환경, SSL 인증 전 CORS 오류 존재, 직접 링크를 작성하여 접속")
+	@Parameter(name = "code", description = "카카오 인가 코드", required = true)
+	@GetMapping("/join")
+	public ResponseEntity<ApiResponse<UserResponse.InfoDto>> joinAuth(KakaoRequest.AuthDto request,
+		HttpServletResponse response) {
+		return ResponseEntity.ok(
+			ApiResponse.success(authCommandService.joinAuth(request.code(), response)));
+	}
+
+	@Operation(summary = "카카오 토큰 재발급")
+	@PostMapping("/reissue")
+	public ResponseEntity<ApiResponse<KakaoResponse.AccessTokenDto>> reissueToken(
+		@CookieValue(name = "refreshToken") KakaoRequest.RefreshTokenDto request, HttpServletResponse response) {
+		return ResponseEntity.ok(
+			ApiResponse.success(authCommandService.reIssueKakaoToken(request.refreshToken(), response)));
+	}
+
+	@Operation(summary = "카카오 로그아웃")
+	@DeleteMapping("/logout")
+	@SecurityRequirement(name = "accessTokenAuth")
+	public ResponseEntity<ApiResponse<KakaoResponse.UserIdDto>> logout(
+		@RequestHeader(ACCESS_TOKEN_HEADER) KakaoRequest.AccessTokenDto request, HttpServletResponse response) {
+		return ResponseEntity.ok(ApiResponse.success(authCommandService.logout(request.accessToken(), response)));
+	}
+
+}
```
@Operation으로 나중에 Swagger에서도 참고하기 편할거 같습니다! 고생하셨습니다

YoungjaeRo commented on src/main/java/com/adit/backend/domain/auth/service/command/AuthCommandService.java:
```diff
@@ -0,0 +1,206 @@
+@Service
+@RequiredArgsConstructor(access = AccessLevel.PROTECTED)
+public class AuthCommandService {
+	public static final String ACCESS_TOKEN_HEADER = "Authorization";
+	public static final String GRANT_TYPE_AUTH_CODE = "authorization_code";
+	public static final String GRANT_TYPE_REFRESH = "refresh_token";
+	public static final String RESPONSE_TYPE = "code";
```
LGTM

YoungjaeRo commented on src/main/java/com/adit/backend/domain/auth/service/command/TokenCommandService.java:
```diff
@@ -0,0 +1,53 @@
+@Service
+@Transactional
+@RequiredArgsConstructor(access = AccessLevel.PROTECTED)
+public class TokenCommandService {
+
+	private final TokenRepository tokenRepository;
+
+	public Token saveOrUpdateToken(KakaoResponse.TokenInfoDto response) {
+		Token token = validateAccessToken(response);
+		log.info("[토큰 저장 완료] accessToken : {} , refreshToken : {}", token.getAccessToken(), token.getRefreshToken());
+		return tokenRepository.save(token);
+	}
+
+	public Token updateTokenEntity(KakaoResponse.AccessTokenDto response, String refreshToken) {
+		Token token = validateRefreshToken(refreshToken);
+		token.updateAccessToken(response.accessToken(), response.expiresIn());
+		log.info("[토큰 갱신 완료] accessToken : {} , refreshToken : {}", token.getAccessToken(), token.getRefreshToken());
+		return tokenRepository.save(token);
+	}
+
+	public void deleteToken(String accessToken) {
+		tokenRepository.deleteByAccessToken(accessToken);
+		log.info("[토큰 삭제 완료] accessToken : {}", accessToken);
+		tokenRepository.flush();
+	}
+
+	private Token validateAccessToken(KakaoResponse.TokenInfoDto response) {
+		return tokenRepository.findByAccessToken(response.accessToken())
+			.orElseGet(response::toEntity);
+	}
+
+	private Token validateRefreshToken(String refreshToken) {
+		return tokenRepository.findTokenByRefreshToken(refreshToken)
+			.orElseThrow(() -> new TokenException(TOKEN_NOT_FOUND));
+	}
+}
```
토큰 관련 비즈니스 로직이 잘 구현되어 있는 거 같습니다. 고생하셨습니다

YoungjaeRo commented on src/main/java/com/adit/backend/domain/user/enums/Role.java:
```diff
@@ -0,0 +1,15 @@
+
+@Getter
+@RequiredArgsConstructor(access = AccessLevel.PROTECTED)
+public enum Role {
+	GUEST("ROLE_GUEST"),
+	USER("ROLE_USER"),
+	ADMIN("ROLE_ADMIN");
+
+	private final String key;
+}
```
굿굿!

YoungjaeRo commented on src/main/java/com/adit/backend/global/config/CorsConfig.java:
```diff
@@ -0,0 +1,31 @@
+import org.springframework.web.filter.CorsFilter;
+
+@Configuration
+public class CorsConfig {
+
+	@Bean
+	public CorsFilter corsFilter() {
+		CorsConfiguration config = new CorsConfiguration();
+
+		config.setAllowCredentials(true);
+		config.addAllowedOrigin("http://localhost:3000");
+		config.addAllowedHeader("*");
+		config.addAllowedMethod("*");
+		config.setExposedHeaders(List.of("accessToken"));
+		config.addExposedHeader("Set-Cookie");
+		config.setMaxAge(3600L);
+
+		UrlBasedCorsConfigurationSource source = new UrlBasedCorsConfigurationSource();
+		source.registerCorsConfiguration("/**", config);
+
+		return new CorsFilter(source);
+	}
+}
```
다음엔 클래스 앞에 주석으로 설명을 달아주시면, 이해하기 더 편할거 같아요!

supsup-hae replied:
네 알겠습니다!

YoungjaeRo commented on src/main/java/com/adit/backend/global/error/GlobalErrorCode.java:
```diff
@@ -0,0 +1,132 @@
+import lombok.Getter;
+import lombok.RequiredArgsConstructor;
+
+/**
+ * [공통 코드] API 통신에 대한 '에러 코드'를 Enum 형태로 관리를 한다.
+ * Global Error CodeList : 전역으로 발생하는 에러코드를 관리한다.
+ * Custom Error CodeList : 업무 페이지에서 발생하는 에러코드를 관리한다
+ * Error Code Constructor : 에러코드를 직접적으로 사용하기 위한 생성자를 구성한다.
+ *
+ * @author lee
+ */
```
좋습니다 ㅎㅎ
